src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+            # starting point.
+            region_name = self.root.invoke.name
+            from psyclone.psyGen import Kern
+            kerns = self.walk(Kern)
+            if len(kerns) == 1:
+                # This PSyData region only has one kernel within it,
+                # so append the kernel name.
+                region_name += ":{0}".format(kerns[0].name)
+            # Add a region index to ensure uniqueness when there are
+            # multiple regions in an invoke.
+            psy_data_nodes = self.root.walk(PSyDataNode)
+            idx = psy_data_nodes.index(self)
+            region_name += ":r{0}".format(idx)
+
+        if not options:
+            options = {}
+
+        pre_variable_list = options.get("pre-var-list", [])
+        post_variable_list = options.get("post-var-list", [])
+        pre_suffix = options.get("pre-var-postfix", "")
+        post_suffix = options.get("post-var-postfix", "")
+
+        # Note that adding a use statement makes sure it is only
+        # added once, so we don't need to test this here!
+        use = UseGen(parent, self.fortran_module, only=True,
+                     funcnames=PSyDataNode.symbols)
+        parent.add(use)
+        var_decl = TypeDeclGen(parent, datatype="PSyDataType",
+                               entity_decls=[self._var_name],
+                               save=True)
+        parent.add(var_decl)
+
+        self._add_call("PreStart", parent,
+                       ["\"{0}\"".format(module_name),
+                        "\"{0}\"".format(region_name),
+                        len(pre_variable_list),
+                        len(post_variable_list)])
+        self.set_region_identifier(module_name, region_name)
+        has_var = pre_variable_list or post_variable_list
+
+        # All variable names get a postfix of "-post" or "-pr".
+        # This enables the netcdf file to store variables that
+        # read read-write (i.e. pre and post), and also avoids
+        # potential name clashes (e.g. when using the PSyData
+        # interface to create a netcdf file), since a Fortran
+        # variable name cannot contain a "-".
+        if has_var:
+            for var_name in pre_variable_list:
+                self._add_call("PreDeclareVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, pre_suffix),
+                                "{0}".format(var_name)])
+            for var_name in post_variable_list:
+                self._add_call("PreDeclareVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, post_suffix),
+                                "{0}".format(var_name)])
+
+            self._add_call("PreEndDeclaration", parent)
+
+            for var_name in pre_variable_list:
+                self._add_call("ProvideVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, pre_suffix),
+                                "{0}".format(var_name)])
```

Comment:
var_name is a str so you don't need this "..".format() I think?

Reply:
Indeed - that was embarrassing (though I did spot a few other ``"{0}".format(...)`` statements elsewhere ;)